File: web/src/lib/sysdeck/bridge/sensors.ts
// SysDeck bridge — sensors (hardware sensor readings, all live)
// Port of bridge/sensors.py: the cockpit edition shells out to
// `sensors -j` (lm-sensors). The web edition does exactly the same:
//   1. `sensors -j` when lm-sensors is installed (the canonical source,
//      chip labels and all — identical to the cockpit panel)
//   2. otherwise it reads the same sysfs sources lm-sensors itself
//      reads (/sys/class/hwmon/hwmon*/*_input with *_label + *_crit,
//      plus /sys/class/thermal/thermal_zone*)
// No fabricated rows, ever: a host with no sensors gets an honest
// empty inventory with a note, not a made-up chip set.
import { readdir } from 'fs/promises'
import { ok, run, which, cached, readText } from './shared'

interface SensorReading {
  label: string
  value: number
  unit: string
  critical?: number
}

interface SensorAdapter {
  name: string
  kind: 'temp' | 'fan' | 'voltage'
  readings: SensorReading[]
}

// ── `sensors -j` (lm-sensors canonical JSON) ─────────────────────────

interface SensorsJson {
  [chip: string]: {
    Adapter?: string
    [key: string]: unknown
  }
}

// lm-sensors JSON shape (real `sensors -j` output):
// { "coretemp-isa-0000": { "Adapter": "ISA adapter",
//     "Package id 0": { "temp1_input": 45, "temp1_max": 100, "temp1_crit": 100 },
//     "Core 0":       { "temp2_input": 44 } },
//   "nct6775-isa-0a40": { "fan1": { "fan1_input": 1240 }, "in0": { "in0_input": 1.344 } } }
function sensorsJsonAdapters(stdout: string): SensorAdapter[] {
  let parsed: SensorsJson
  try {
    parsed = JSON.parse(stdout) as SensorsJson
  } catch {
    return []
  }
  const out: SensorAdapter[] = []
  for (const [chip, fields] of Object.entries(parsed)) {
    const temps: SensorReading[] = []
    const fans: SensorReading[] = []
    const volts: SensorReading[] = []
    for (const [label, raw] of Object.entries(fields)) {
      if (label === 'Adapter' || typeof raw !== 'object' || raw === null) continue
      const entry = raw as Record<string, number>
      // the feature prefix in the *_input key is the authoritative kind
      const inputKey = Object.keys(entry).find((k) => /^(temp|fan|in)\d+_input$/.test(k))
      if (!inputKey) continue
      const value = entry[inputKey]
      if (typeof value !== 'number' || !Number.isFinite(value)) continue
      const prefix = inputKey.replace(/_input$/, '')
      if (inputKey.startsWith('temp')) {
        const crit = entry[`${prefix}_crit`] ?? entry[`${prefix}_max`] ?? null
        temps.push({
          label,
          value: Math.round(value * 10) / 10,
          unit: '°C',
          ...(typeof crit === 'number' && Number.isFinite(crit) ? { critical: Math.round(crit * 10) / 10 } : {}),
        })
      } else if (inputKey.startsWith('fan')) {
        fans.push({ label, value: Math.round(value), unit: 'RPM' })
      } else {
        volts.push({ label, value: Math.round(value * 1000) / 1000, unit: 'V' })
      }
    }
    if (temps.length) out.push({ name: chip, kind: 'temp', readings: temps })
    if (fans.length) out.push({ name: chip, kind: 'fan', readings: fans })
    if (volts.length) out.push({ name: chip, kind: 'voltage', readings: volts })
  }
  return out
}

// ── sysfs collectors (what lm-sensors reads under the hood) ──────────
// All reads are async — the hwmon walk never blocks the event loop.

async function readNum(path: string): Promise<number | null> {
  const t = await readText(path)
  if (!t) return null
  const v = Number(t.trim())
  return Number.isFinite(v) ? v : null
}

async function readStr(path: string): Promise<string | null> {
  const t = await readText(path)
  return t.trim() || null
}

async function sysfsAdapters(): Promise<SensorAdapter[]> {
  const out: SensorAdapter[] = []
  let hwmons: string[] = []
  try {
    hwmons = await readdir('/sys/class/hwmon')
  } catch {
    return out // /sys/class/hwmon absent — honest empty
  }
  for (const h of hwmons) {
    const base = `/sys/class/hwmon/${h}`
    const name = (await readStr(`${base}/name`)) ?? h
    let files: string[] = []
    try {
      files = await readdir(base)
    } catch {
      continue
    }
    const temps: SensorReading[] = []
    const fans: SensorReading[] = []
    const volts: SensorReading[] = []
    for (const f of files) {
      const m = f.match(/^(temp|fan|in)(\d+)_input$/)
      if (!m) continue
      const raw = await readNum(`${base}/${f}`)
      if (raw === null) continue
      const label = (await readStr(`${base}/${m[1]}${m[2]}_label`)) ?? `${m[1]}${m[2]}`
      if (m[1] === 'temp') {
        const crit = await readNum(`${base}/${m[1]}${m[2]}_crit`)
        temps.push({
          label,
          value: Math.round((raw / 1000) * 10) / 10,
          unit: '°C',
          ...(crit !== null ? { critical: Math.round((crit / 1000) * 10) / 10 } : {}),
        })
      } else if (m[1] === 'fan') {
        fans.push({ label, value: raw, unit: 'RPM' })
      } else {
        volts.push({ label, value: Math.round((raw / 1000) * 100) / 100, unit: 'V' })
      }
    }
    if (temps.length) out.push({ name, kind: 'temp', readings: temps })
    if (fans.length) out.push({ name, kind: 'fan', readings: fans })
    if (volts.length) out.push({ name, kind: 'voltage', readings: volts })
  }
  return out
}

async function thermalZones(): Promise<SensorAdapter[]> {
  const out: SensorAdapter[] = []
  let zones: string[] = []
  try {
    zones = await readdir('/sys/class/thermal')
  } catch {
    return out // /sys/class/thermal absent — honest empty
  }
  for (const z of zones) {
    if (!/^thermal_zone\d+$/.test(z)) continue
    const temp = await readNum(`/sys/class/thermal/${z}/temp`)
    if (temp === null) continue
    const type = (await readStr(`/sys/class/thermal/${z}/type`)) ?? z
    out.push({
      name: `${type} (thermal_zone)`,
      kind: 'temp',
      readings: [{ label: type, value: Math.round((temp / 1000) * 10) / 10, unit: '°C' }],
    })
  }
  return out
}

// ── collection ───────────────────────────────────────────────────────

async function collectSysfs(): Promise<{ adapters: SensorAdapter[]; note?: string }> {
  // sysfs fallback rung — always available, zero dependencies
  const adapters = [...(await sysfsAdapters()), ...(await thermalZones())]
  if (adapters.length === 0) {
    return {
      adapters,
      note: 'no sensor readings on this host (probed /sys/class/hwmon + /sys/class/thermal; install lm-sensors and run sensors-detect for chip-level labels)',
    }
  }
  return { adapters }
}

/** Sensor chain: `sensors -j` (lm-sensors, the cockpit panel's source)
 *  → raw sysfs hwmon + thermal zones → honest empty. The whole sweep is
 *  TTL-cached so the summary/temps/fans/voltages polls share one pass. */
async function collect(): Promise<{ adapters: SensorAdapter[]; note?: string }> {
  return cached('sensors:collect', 3000, async () => {
    if (await which('sensors')) {
      const r = await run('sensors', ['-j'], 10_000)
      if (r.rc === 0 && r.stdout.trim().startsWith('{')) {
        const adapters = sensorsJsonAdapters(r.stdout)
        if (adapters.length > 0) {
          return {
            adapters,
            note: 'real `sensors -j` (lm-sensors) — the same source the cockpit edition reads',
          }
        }
        // lm-sensors present but no chips configured → step down to sysfs
        // and say so honestly
        const sys = await collectSysfs()
        return { ...sys, note: `sensors -j returned no chips (run sensors-detect); showing raw sysfs readings — ${sys.note ?? 'direct /sys/class/hwmon + thermal_zone reads'}` }
      }
    }
    return collectSysfs()
  })
}

// ── commands ─────────────────────────────────────────────────────────

export const commands = {
  summary: async () => {
    const { adapters, note } = await collect()
    return ok({ adapters, source: 'live' }, 'live', note)
  },

  temps: async () => {
    const { adapters, note } = await collect()
    return ok(
      { adapters: adapters.filter((a) => a.kind === 'temp'), source: 'live' },
      'live',
      note,
    )
  },

  fans: async () => {
    const { adapters, note } = await collect()
    return ok(
      { adapters: adapters.filter((a) => a.kind === 'fan'), source: 'live' },
      'live',
      note,
    )
  },

  voltages: async () => {
    const { adapters, note } = await collect()
    return ok(
      { adapters: adapters.filter((a) => a.kind === 'voltage'), source: 'live' },
      'live',
      note,
    )
  },
}

File: bridge/sensors.py
#!/usr/bin/env python3
"""
SysDeck - Sensors Bridge Helper
Author: Jeremy Anderson (https://dcos.net)

Aggregates hardware sensor data from lm_sensors (`sensors -j`) and
the cockpit-sensors plugin (https://github.com/ocristopfer/cockpit-sensors)
into a structured JSON document.

cockpit-sensors is MIT licensed by ocristopfer. This bridge helper invokes
sensors as a separate process via subprocess — the suite (MIT) and
lm_sensors remain independent programs. No cockpit-sensors code is bundled;
the suite's panel provides its own rendering of the sensors data.

Sensor chain (same step-down the web bridge runs):
    `sensors -j` (lm-sensors)  →  raw /sys/class/hwmon + thermal zones  →
    honest empty. lm-sensors is the preferred rung (chip-level labels);
    sysfs is the dependency-free rung; a host with no readable sensors
    returns an empty document, never fabricated readings.

Usage:
    python3 -m sysdeck.bridge.sensors summary
    python3 -m sysdeck.bridge.sensors temps
    python3 -m sysdeck.bridge.sensors fans
    python3 -m sysdeck.bridge.sensors voltages
"""

import json
import re
import shutil
import subprocess
import sys
from pathlib import Path
from typing import Any


SENSORS_LICENSE = "MIT (lm_sensors) + MIT (cockpit-sensors)"
SENSORS_AUTHOR = "ocristopfer (cockpit-sensors), lm_sensors project"
SENSORS_URL = "https://github.com/ocristopfer/cockpit-sensors"

# Scrubbed child environment: parsed output stays locale-stable and no
# console process state leaks into children.
SCRUBBED_ENV = {"PATH": "/usr/sbin:/usr/bin:/sbin:/bin", "LANG": "C", "LC_ALL": "C"}


def run_sensors(args: list[str]) -> str | None:
    """Run sensors with the given args, returning stdout; None when the
    binary is absent or fails (the caller steps down to sysfs)."""
    if not shutil.which("sensors"):
        return None
    try:
        r = subprocess.run(
            ["sensors", *args], capture_output=True, text=True, check=False,
            timeout=10, env=SCRUBBED_ENV,
        )
    except (subprocess.TimeoutExpired, OSError):
        return None
    return r.stdout if r.returncode == 0 else None


# ── sysfs rung (what lm-sensors reads under the hood) ────────────────

_INPUT_RE = re.compile(r"^(temp|fan|in)(\d+)_input$")


def _read_str(path: Path) -> str | None:
    try:
        return path.read_text().strip() or None
    except OSError:
        return None


def _read_num(path: Path) -> float | None:
    raw = _read_str(path)
    if raw is None:
        return None
    try:
        return float(raw)
    except ValueError:
        return None


def sysfs_adapters() -> dict[str, dict[str, Any]]:
    """Raw /sys/class/hwmon readings in the sensors -j shape:
    { adapter: { sensor_key: {"input": v, ...} } }."""
    out: dict[str, dict[str, Any]] = {}
    hwmon_root = Path("/sys/class/hwmon")
    try:
        chips = sorted(hwmon_root.iterdir())
    except OSError:
        return out
    for chip in chips:
        name = _read_str(chip / "name") or chip.name
        adapter: dict[str, Any] = {}
        try:
            files = sorted(f.name for f in chip.iterdir())
        except OSError:
            continue
        for fname in files:
            m = _INPUT_RE.match(fname)
            if not m:
                continue
            kind, idx = m.group(1), m.group(2)
            raw = _read_num(chip / fname)
            if raw is None:
                continue
            label = _read_str(chip / f"{kind}{idx}_label") or f"{kind}{idx}"
            entry: dict[str, Any] = {}
            if kind == "temp":
                entry["temp1_input"] = round(raw / 1000.0, 1)
                crit = _read_num(chip / f"{kind}{idx}_crit")
                if crit is not None:
                    entry["temp1_crit"] = round(crit / 1000.0, 1)
            elif kind == "fan":
                entry["fan1_input"] = int(raw)
            else:
                entry["in0_input"] = round(raw / 1000.0, 3)
            adapter[label] = entry
        if adapter:
            out[name] = adapter
    return out


def thermal_zones() -> dict[str, dict[str, Any]]:
    """ACPI thermal zones as temperature adapters."""
    out: dict[str, dict[str, Any]] = {}
    root = Path("/sys/class/thermal")
    try:
        zones = sorted(z for z in root.iterdir() if z.name.startswith("thermal_zone"))
    except OSError:
        return out
    for z in zones:
        temp = _read_num(z / "temp")
        if temp is None:
            continue
        ztype = _read_str(z / "type") or z.name
        out[f"{ztype} (thermal_zone)"] = {ztype: {"temp1_input": round(temp / 1000.0, 1)}}
    return out


def summary() -> dict[str, Any]:
    """Full sensor summary.

    Chain: `sensors -j` → sysfs hwmon + thermal zones → honest empty.
    Returns a nested dict: { adapter: { sensor: { field: value } } }"""
    output = run_sensors(["-j"])
    if output is not None:
        try:
            data = json.loads(output)
            if data:
                return data
            # sensors present but no chips configured → sysfs rung, labeled
            sysfs = {**sysfs_adapters(), **thermal_zones()}
            return {"_note": "sensors -j returned no chips (run sensors-detect); showing raw sysfs readings", **sysfs}
        except json.JSONDecodeError:
            pass
    return {**sysfs_adapters(), **thermal_zones()}


def _filtered(predicate) -> dict[str, Any]:
    data = summary()
    result: dict[str, Any] = {}
    for adapter, sensors in data.items():
        filtered = {key: val for key, val in sensors.items() if predicate(key)}
        if filtered:
            result[adapter] = filtered
    return result


def temps() -> dict[str, Any]:
    """Temperature sensors only — filter for keys containing 'temp' or 'Core'."""
    return _filtered(
        lambda key: "temp" in key.lower() or "core" in key.lower() or "tctl" in key.lower()
    )


def fans() -> dict[str, Any]:
    """Fan speed sensors only — filter for keys containing 'fan'."""
    return _filtered(lambda key: "fan" in key.lower())


def voltages() -> dict[str, Any]:
    """Voltage sensors only — filter for keys containing 'in' (lm_sensors convention)."""
    return _filtered(
        lambda key: key.startswith("in") or "vcore" in key.lower() or "vbat" in key.lower()
    )


COMMANDS = {
    "summary": lambda _args: summary(),
    "temps": lambda _args: temps(),
    "fans": lambda _args: fans(),
    "voltages": lambda _args: voltages(),
}


def main(argv: list[str]) -> int:
    if not argv or argv[0] in ("-h", "--help"):
        print(__doc__)
        return 0
    cmd = COMMANDS.get(argv[0])
    if not cmd:
        print(f"Unknown subcommand: {argv[0]}", file=sys.stderr)
        return 2
    print(json.dumps(cmd(argv[1:]), indent=2))
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))
